src/lib/common/string.cpp

```diff
@@ -922,122 +922,52 @@ bool str2double(const char* s, double* dP)
 }
 
 
-
 /* ****************************************************************************
 *
-* decimalDigits
-*
-* This function counts the number of decimal digits of a given float, to a maximum of
-* PRECISION_DIGITS. The algorithm is inspired in http://stackoverflow.com/a/1083316/1485926
-* but with a "cutting condition" needed due to float representation may have an infinite
-* number of decimals, e.g. 3.14 could be internally coded as 3.1399999.
-*
-* FIXME #2425: this function is not perfect and could be improved. For example,
-* considering the following
-*
-*   "A1":  42.9,
-*   "A2":  42.99,
-*   "A3":  42.999,
-*   "A4":  42.9999,
-*   "A5":  42.99999,
-*   "A6":  42.999999,
-*   "A7":  42.9999999,
-*   "A8":  42.99999999,
-*   "A9":  42.999999999,
-*   "A10": 42.9999999999,,
-*
-*   "A1":  42.1,
-*   "A2":  42.01,
-*   "A3":  42.001,
-*   "A4":  42.0001,
-*   "A5":  42.00001,
-*   "A6":  42.000001,
-*   "A7":  42.0000001,
-*   "A8":  42.00000001,
-*   "A9":  42.000000001,
-*   "A10": 42.0000000001,
-*
-* what we get is:
-*
-*   "A1":  42.9,
-*   "A2":  42.99,
-*   "A3":  42.999,
-*   "A4":  42.9999,
-*   "A5":  42.99999,
-*   "A6":  42.999999000, (fail)
-*   "A7":  42.999999900, (fail)
-*   "A8":  42.999999990, (fail)
-*   "A9":  42.999999999,
-*   "A10": 43.000000000, (fail, although probably not due to this function but the caller)
-*
-*   "A1":  42.1,
-*   "A2":  42.01,
-*   "A3":  42.001,
-*   "A4":  42.0001,
-*   "A5":  42.00001,
-*   "A6":  42.000001000, (fail)
-*   "A7":  42.000000100, (fail)
-*   "A8":  42.000000010, (fail)
-*   "A9":  42,           (fail)
-*   "A10": 42,
-*
+* toString - 
 */
-unsigned int decimalDigits(double d)
+char* toString(double number, char* buf, int bufSize)
```

Comment:
Instead of defining a new function, this algorithm can be included in the current toString(double) function. I mean:

```
template <> std::string toString(double f)
{
  char  buf[STRING_SIZE_FOR_DOUBLE];
  int bufSize = sizeof(buf);

  long long  intPart   = (long long) f;
  double     diff      = f - intPart;
  bool       isInteger = false;

  // abs value for 'diff'
  diff = (diff < 0)? -diff : diff;

  if (diff > 0.9999999998)
  {
    intPart += 1;
    isInteger = true;
  }
  else if (diff < 0.000000001)  // it is considered an integer
  {
    isInteger = true;
  }

  if (isInteger)
  {
    snprintf(buf, bufSize, "%lld", intPart);
  }
  else
  {
    snprintf(buf, bufSize, "%.9f", f);

    // Clear out any unwanted trailing zeroes
    char* last = &buf[strlen(buf) - 1];

    while ((*last == '0') && (last != buf))
    {
      *last = 0;
      --last;
    }
  }

  return std::string(buf);
}
```

This way no additional modifications in other places of the code are required as the calling interface for toString(double) doesn't change.

Reply:
Let's try to avoid unnecessarily slow code, and try to avoid returning objects on the stack.

About changing the calls to `toString`, those changes are already made.
There is absolutely nothing to win in changing the function from returning a `std::string` instead of the `char*` it is returning right now.  That would only give more work and slower code.

An additional advantage with the current implementation - it can be used without assignment on the caller side:
```
char buf[xxx];
toString(dValue, buf, sizeof(buf));
```

NTC

Reply:
Let’s try to avoid unnecessary changes not directly related with the issue the PR is attempting to solve, especially when these changes increase the complexity of the code.

Complexity increase should be evident. For instance, the statement

```
return toString(numberValue);
```

is much simpler than

```
char numberBuffer[STRING_SIZE_FOR_DOUBLE];
return toString(numberValue, numberBuffer, sizeof(numberBuffer));
```

The need of to pre-declare an extra variable increases complexity. Change function call signature from 1 to 3 parameters increase complexity. Having the same parameter (numberBuffer) at the same time as input parameter and return parameter increases complexity. And multiply this increase of complexity for all places in which the function is being called. Keeping complexity low is a win.

The proposed solution (i.e. apply the rounding algorithm in the toString() function but without changing its signature) doesn’t give *slower* code. The code is not *slower than* the one currently in master. In fact, it is probably faster (as std::ostringstream is no longer used). The size of std::string in the stack is not a problem (as, as we measure, the size is exact the same than the one of a char* for the reference Orion sytem, i.e. RHEL/CentOS 7 and, even in other system, the increase is not significant).

This PR should stick to the issue is attempting to solve and not trying to change a function pattern in a way it has not been previously agreed. Before implementing this kind of changes we should have the proper discussion (maybe the right place to have it is #1298).

Reply:
There is really not much to discuss here, but sure, we can create an issue and discuss it there.
But right now we have more urgent matters.
This PR fixes a pretty ugly bug that destroys big numeric values and people are waiting for it to be fixed.
We need to merge this PR ASAP.
After that we have all the time in the world to discuss this matter.

So again,

NTC